Allow FSDP to have ignored modules out of wrapped root

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91079

Motivations for this change:

1. TorchRec returns inconsistent results on `m.named_parameters()`
   and `m.m1.named_parameters()` if m1 is a `ShardedModule`. Basically,
   `ShardedModule` appears in `m.named_modules()`, but its parameters
   are not in `m.named_parameters()`. As a result, when we identify
   `ShardedModule` and pass them as `ignored_modules` to FSDP, FSDP
   complains about key error in `_get_ignored_params`.
2. If users are manually wrapping submodules with FSDP, it could be
   easier for them to keep a global set of ignored parameters, instead
   of create a new collection for every FSDP invocation.

Given the above two reasons, we allow FSDP to have ignored modules
out of the wrapped root module.

Differential Revision: [D42132394](https://our.internmc.facebook.com/intern/diff/D42132394)

mrshenli committed:
Fix FSDP with ignored modules out of wrapped root

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_ignored_modules.py:
```diff
@@ -6,6 +6,7 @@
 import torch.nn as nn
 from torch import distributed as dist
 from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp.api import ShardingStrategy
```
nit: Let us just import from `.fsdp` for now even though `.fsdp.api` works.
```suggestion
from torch.distributed.fsdp import ShardingStrategy
```

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -570,23 +570,36 @@ def _get_ignored_params(
+            # This is useful when user apply FSDP manually to different
+            # submodules with the same global set of ignored parameters.
+            warnings.warn(
+                f"Paramter {name} is in the ignored modules passed to FSDP, "
```
nit:
```suggestion
                f"Parameter {name} is in the ignored modules passed to FSDP, "
```

awgu commented on test/distributed/fsdp/test_fsdp_ignored_modules.py:
```diff
@@ -225,6 +227,23 @@ def test_diff_ignored_modules_across_ranks(
+        ignored_modules = list(model.layer1.children())[1:]
+        model.layer1 = FSDP(
+            model.layer1,
+            sharding_strategy=ShardingStrategy.SHARD_GRAD_OP,
```
I am wondering if you think there is any benefit to leave a comment somewhere in this test to say that sharding strategy should not matter or at least some comment explaining the choice of sharding strategy (e.g. `model.layer1` vs. `model.layer3`).

awgu replied:
I am also wondering if you think it would be worth adding a comment to highlight how the ignored modules for the second application of FSDP to `model.layer3` is where the new code path is being tested, i.e. that there are ignored modules passed to L241 that are not in `model.layer3`'s tree.

mrshenli replied:
Hey @awgu, thanks for the quick stamp. Let me add those comments.

awgu reviewed (approved):
The semantics of this change make sense to me, and the PR looks good overall! I just left some minor comments.

mrshenli committed:
Update on "Allow FSDP to have ignored modules out of wrapped root"


Motivations for this change:

1. TorchRec returns inconsistent results on `m.named_parameters()`
   and `m.m1.named_parameters()` if m1 is a `ShardedModule`. Basically,
   `ShardedModule` appears in `m.named_modules()`, but its parameters
   are not in `m.named_parameters()`. As a result, when we identify
   `ShardedModule` and pass them as `ignored_modules` to FSDP, FSDP
   complains about key error in `_get_ignored_params`.
2. If users are manually wrapping submodules with FSDP, it could be
   easier for them to keep a global set of ignored parameters, instead
   of create a new collection for every FSDP invocation.

Given the above two reasons, we allow FSDP to have ignored modules
out of the wrapped root module.

[ghstack-poisoned]

mrshenli committed:
Update on "Allow FSDP to have ignored modules out of wrapped root"


Motivations for this change:

1. TorchRec returns inconsistent results on `m.named_parameters()`
   and `m.m1.named_parameters()` if m1 is a `ShardedModule`. Basically,
   `ShardedModule` appears in `m.named_modules()`, but its parameters
   are not in `m.named_parameters()`. As a result, when we identify
   `ShardedModule` and pass them as `ignored_modules` to FSDP, FSDP
   complains about key error in `_get_ignored_params`.
2. If users are manually wrapping submodules with FSDP, it could be
   easier for them to keep a global set of ignored parameters, instead
   of create a new collection for every FSDP invocation.

Given the above two reasons, we allow FSDP to have ignored modules
out of the wrapped root module.

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42132394).

mrshenli commented:
@pytorchbot merge -g

mrshenli commented:
@pytorchbot merge -f "no test failure, but I still cannot merge with -g"

rohan-varma reviewed (commented):
LGTM, thanks!
